hedera-mirror-rest/api/v1/openapi.yml

```
@@ -3671,14 +3677,33 @@ components:
       example: cryptotransfer
       schema:
         $ref: '#/components/schemas/TransactionTypes'
+    transactionIdOrHashPathParam:
+      name: transactionIdOrHash
+      in: path
+      required: true
+      description: Transaction id or hash
+      examples:
+        transactionId:
+          value: 0.0.10-1234567890-000000000
+        transactionHashInBase64url:
+          value: rovr8cn6DzCTVuSAV_YEevfN5jA30FCdFt3Dsg4IUVi_3xTRU0XBsYsZm3L-1Kxv
+        transactionHashInHex:
+          value: ab4af784ae69ca3e1d17fef8491f7f89a0e8a3b80ad3748b841110f0345ada53456bc14bbe9ee6f441829c2849933ac4
+        transactionHashInHexWithPrefix:
+          value: '0xab4af784ae69ca3e1d17fef8491f7f89a0e8a3b80ad3748b841110f0345ada53456bc14bbe9ee6f441829c2849933ac4'
+      schema:
+        oneOf:
+          - $ref: '#/components/schemas/HederaHashInBase64url'
+          - $ref: '#/components/schemas/HederaHashInHex'
+          - $ref: '#/components/schemas/TransactionIdStr'
```

While this is reusable from a yaml perspective, it does not generate actual working code since there are three types. The generator doesn't know how to create an API client for this and will end up generating a wrapper type that doesn't exist.

Going forward, we need to have a single schema definition for any parameter. That single schema can still be reused elsewhere. So here this would be of type string with a single regex with an OR like `^(...|...|...)$`.

```suggestion
      schema:
        $ref: '#/components/schemas/TransactionIdOrHash'
```

---

I don't see the issue of non-existing wrapper type. Using the `oneOf` here, the openapi generator will generates `GetTransactionByIdOrHashTransactionIdOrHashParameter.java` which indicates it's an `oneOf` and the accepted schema is `String`. That makes sense since all of 3 possible schemas are strings.

---

Except it doesn't generate an actual `GetTransactionByIdOrHashTransactionIdOrHashParameter.java`

---

It does. btw, I made local changes to add `or hash` to the openapi where it applies so my generated class name will be:

```shell
$ find -type f -name 'GetTransactionByIdOrHash*.java'
./model/GetTransactionByIdOrHashTransactionIdOrHashParameter.java

$ grep -n 'GetTransactionByIdOrHashTransactionIdOrHashParameter' api/TransactionsApi.java
252:    public okhttp3.Call getTransactionByIdOrHashCall(GetTransactionByIdOrHashTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled, final ApiCallback _callback) throws ApiException {
307:    private okhttp3.Call getTransactionByIdOrHashValidateBeforeCall(GetTransactionByIdOrHashTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled, final ApiCallback _callback) throws ApiException {
336:    public TransactionByIdResponse getTransactionByIdOrHash(GetTransactionByIdOrHashTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled) throws ApiException {
357:    public ApiResponse<TransactionByIdResponse> getTransactionByIdOrHashWithHttpInfo(GetTransactionByIdOrHashTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled) throws ApiException {
380:    public okhttp3.Call getTransactionByIdOrHashAsync(GetTransactionByIdOrHashTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled, final ApiCallback<TransactionByIdResponse> _callback) throws ApiException {
```

---

Well, I wasn't sure of the exact error since I didn't want to have to dig up my gradle branch. But since you're inquiring, I re-ran it against your `openapi.yml`. This is the error while attempting to compile the API:

```java
/Users/steven/projects/hedera/hedera-mirror-node/hedera-mirror-monitor/build/generate-resources/main/src/main/java/com/hedera/mirror/rest/api/TransactionsApi.java:7: error: cannot find symbol
import com.hedera.mirror.rest.model.OneOfstringstringstring;
                                   ^
  symbol:   class OneOfstringstringstring
  location: package com.hedera.mirror.rest.model
/Users/steven/projects/hedera/hedera-mirror-node/hedera-mirror-monitor/build/generate-resources/main/src/main/java/com/hedera/mirror/rest/api/TransactionsApi.java:133: error: cannot find symbol
    private ResponseSpec getTransactionByIdRequestCreation(GetTransactionByIdTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled) throws WebClientResponseException {
                                                           ^
  symbol:   class GetTransactionByIdTransactionIdOrHashParameter
  location: class TransactionsApi
/Users/steven/projects/hedera/hedera-mirror-node/hedera-mirror-monitor/build/generate-resources/main/src/main/java/com/hedera/mirror/rest/api/TransactionsApi.java:177: error: cannot find symbol
    public Mono<TransactionByIdResponse> getTransactionById(GetTransactionByIdTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled) throws WebClientResponseException {
                                                            ^
  symbol:   class GetTransactionByIdTransactionIdOrHashParameter
  location: class TransactionsApi
/Users/steven/projects/hedera/hedera-mirror-node/hedera-mirror-monitor/build/generate-resources/main/src/main/java/com/hedera/mirror/rest/api/TransactionsApi.java:182: error: cannot find symbol
    public Mono<ResponseEntity<TransactionByIdResponse>> getTransactionByIdWithHttpInfo(GetTransactionByIdTransactionIdOrHashParameter transactionIdOrHash, Integer nonce, Boolean scheduled) throws WebClientResponseException {
                                                                                        ^
  symbol:   class GetTransactionByIdTransactionIdOrHashParameter
  location: class TransactionsApi
```

So it's true the `GetTransactionByIdTransactionIdOrHashParameter` class exists, but it's missing an import to it. But it does have an import to `OneOfstringstringstring`, which doesn't exist. So I think it's confusing the two and can't generate a valid API when using oneof.

Even if you were to somehow correct the missing import and remove the bad one, the actual contents of `GetTransactionByIdTransactionIdOrHashParameter` are empty and are useless:

```java
@JsonPropertyOrder({
})
@JsonTypeName("getTransactionById_transactionIdOrHash_parameter")
@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", date = "2022-08-10T10:36:50.387417-05:00[America/Chicago]")
public class GetTransactionByIdTransactionIdOrHashParameter {
  public GetTransactionByIdTransactionIdOrHashParameter() {
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    return true;
  }

  @Override
  public int hashCode() {
    return Objects.hash();
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class GetTransactionByIdTransactionIdOrHashParameter {\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }

}
```

I also ran it against the maven plugin and the same error resulted.

---

In summary, using oneof results in generating code that won't compile and even if it did the wrapper object is useless. Please remove the oneof.